cmd/grpc.ext/grpc.go

```diff
@@ -0,0 +1,137 @@
+	osquery_logger "github.com/kolide/osquery-go/plugin/logger"
+	"github.com/pkg/errors"
+)
+
+func main() {
+	var (
+		flSocketPath = flag.String("socket", "", "")
+		flTimeout    = flag.Int("timeout", 0, "")
+		flVerbose    = flag.Bool("verbose", false, "")
+		_            = flag.Int("interval", 0, "")
+	)
+	flag.Parse()
+
+	timeout := time.Duration(*flTimeout) * time.Second
+
+	// allow for osqueryd to create the socket path
+	time.Sleep(2 * time.Second)
+
+	logger := kolidelog.NewLogger(os.Stderr)
+	if *flVerbose {
+		logger.AllowDebug()
+	}
+
+	client, err := osquery.NewClient(*flSocketPath, timeout)
+	if err != nil {
+		logger.Fatal("err", err, "creating osquery extension client")
+	}
+
+	var (
+		enrollSecret  = env.String("KOLIDE_LAUNCHER_ENROLL_SECRET", "")
+		rootDirectory = env.String("KOLIDE_LAUNCHER_ROOT_DIRECTORY", "")
+
+		serverURL    = env.String("KOLIDE_LAUNCHER_HOSTNAME", "")
+		insecureTLS  = env.Bool("KOLIDE_LAUNCHER_INSECURE", false)
+		insecureGRPC = env.Bool("KOLIDE_LAUNCHER_INSECURE_GRPC", false)
+
+		// TODO(future pr): these values are unset
+		// they'll have to be parsed from a string
+		certPins [][]byte
+		rootPool *x509.CertPool
+	)
+	conn, err := service.DialGRPC(
+		serverURL,
+		insecureTLS,
+		insecureGRPC,
+		certPins,
+		rootPool,
+		logger,
+	)
+	remote := service.New(conn, level.Debug(logger))
+
+	extOpts := grpcext.ExtensionOpts{
+		EnrollSecret:    enrollSecret,
+		Logger:          logger,
+		LoggingInterval: 30 * time.Second,
+	}
+
+	db, err := bolt.Open(filepath.Join(rootDirectory, "launcher.db"), 0600, nil)
+	if err != nil {
+		logger.Fatal("err", errors.Wrap(err, "open local store"))
+	}
+	defer db.Close()
+
+	ext, err := grpcext.NewExtension(remote, db, extOpts)
+	if err != nil {
+		logger.Fatal("err", errors.Wrap(err, "starting grpc extension"))
+	}
+
+	// create an extension server
+	server, err := osquery.NewExtensionManagerServer(
+		"com.kolide.grpc_extension",
+		*flSocketPath,
+		osquery.ServerTimeout(timeout),
+	)
+	if err != nil {
+		logger.Fatal("err", err, "msg", "creating osquery extension server")
+	}
+
+	configPlugin := config.NewPlugin("kolide_grpc", ext.GenerateConfigs)
+	distributedPlugin := distributed.NewPlugin("kolide_grpc", ext.GetQueries, ext.WriteResults)
+	loggerPlugin := osquery_logger.NewPlugin("kolide_grpc", ext.LogString)
+
+	server.RegisterPlugin(configPlugin, distributedPlugin, loggerPlugin)
+
+	ext.SetQuerier(&queryier{client: client})
+	ctx := context.Background()
+	_, invalid, err := ext.Enroll(ctx)
+	if err != nil {
+		logger.Fatal("err", errors.Wrap(err, "enrolling host"))
+	}
+	if invalid {
+		logger.Fatal(errors.Wrap(err, "invalid enroll secret"))
+	}
+	ext.Start()
+	defer ext.Shutdown()
+
+	ext.Start()
+	defer ext.Shutdown()
+
+	if err := server.Run(); err != nil {
+		logger.Fatal("err", err)
+	}
+}
+
+type queryier struct {
+	client *osquery.ExtensionManagerClient
+}
+
+func (q *queryier) Query(query string) ([]map[string]string, error) {
```

Comment:
Copy paste? This doesn't seem to be used.

Reply:
It's used on line 37. It's required to make the RequestEnrollment method add initial enrollment data from the client. 

Launcher has this method on the osquery runner.